### PR TITLE
SQL-1434: Support setting SQL_ATTR_CURRENT_CATALOG

### PR DESCRIPTION
**Pull request by terakilobyte**
Turns out we supported getting this attribute in SQLGetConnectAttrW but never implemented setting it via SQLSetConnectAttrW.

### CODE REVIEW

**Review by bucaojit — APPROVED** [review]
LGTM!  Thanks for getting to this right away 🥇

**Review by pmeredit — APPROVED** [review]
Great! Can't believe we missed this before :P

**Comment by pmeredit** [comment]
Actually hold on a second.

**Review by pmeredit — CHANGES_REQUESTED** [review]
You're going to need to set the current_catalog in the `MongoConnection` struct for this to work right, which exists inside of the `Connection` handle.

In fact, it'd make more sense to remove the current_catalog from the attributes entirely and only report the value from MongoConnection.

HOWEVER, it might be allowable to set the current_catalog before CONNECTING the Connection. In which case we do need to preserve the attribute (and set it in both places if both places exist when Set is called, and also populate the field in the MongoConnection from the attributes when connect is called)

Sorry for this, this seems to be a bug in what we were already doing.

**Comment by terakilobyte** [comment]
@pmeredit This is ready to look at again. The current tests cover the surface area of the changes pretty well (I had to go through with a debugger to remove a threadlock issue), but I feel like maybe I need to add more.

File: core/src/conn.rs [hunk 1]
```diff
@@ -6,24 +6,27 @@ use mongodb::sync::Client;
...
+    // for this Connection.
+    pub current_catalog: Option<String>,
+    // SQL_ATTR_LOGIN_TIMEOUT: SQLUINTEGER, timeout in seconds
+    // to wait for a login request to complete.
+    pub login_timeout: Option<u32>,
+    // SQL_ATTR_CONNECTION_TIMEOUT: SQLUINTER, timeout in seconds
+    // to wait for any operation on a connection to timeout (other than
+    // initial login).
+    pub connection_timeout: Option<u32>,
+}
```
Comment (terakilobyte) [inline core/src/conn.rs:21]: This definition was relocated here and is shared between core and odbc

File: core/src/conn.rs [hunk 2]
```diff
@@ -6,24 +6,27 @@ use mongodb::sync::Client;
...
-    /// We stick with mongo terminology here and ODBC
-    /// terminology in the odbc wrappers, hence
-    /// current_db here and current_catalog in the
-    /// odbc/handles code.
-    pub current_db: Option<String>,
-    /// Number of seconds to wait for any request on the connection to complete before returning to
-    /// the application.
-    /// Comes from SQL_ATTR_CONNECTION_TIMEOUT if set. Used any time there is a time out in a
-    /// situation not associated with query execution or login.
-    pub operation_timeout: Option<Duration>,
```
Comment (terakilobyte) [inline core/src/conn.rs:24]: All removed in favor of attributes

File: odbc/src/api/functions.rs [hunk 3]
```diff
@@ -868,22 +868,33 @@ fn sql_driver_connect(conn: &Connection, odbc_uri_string: &str) -> Result<MongoC
...
+            .as_deref()
+            .map(|s| s.to_string());
+        database = if catalog_database.is_some() {
+            catalog_database
+        } else {
+            odbc_uri.remove(&["database"])
+        };
+    }
+    let mut attrs = conn.attributes.write().unwrap().take().unwrap();
+    attrs.current_catalog = database;
```
Comment (terakilobyte) [inline odbc/src/api/functions.rs:891]: This feels a little verbose, but makes the Rust compiler happy. This thankfully is also the only place that I take a read lock immediately followed by a write lock on the same RwLock, so the only place this extra scope is introduced. I'm thinking about taking one write lock here for efficiency. Thoughts?

File: odbc/src/api/functions.rs [hunk 4]
```diff
@@ -1321,25 +1351,35 @@ unsafe fn sql_get_connect_attrw_helper(
...
+
+        let has_connection = conn.has_mongo_connection();
+        // unfortunate clones here, but the types in a ConnectionAttributes struct are small
+        let attributes = if has_connection {
+            mongo_connection.as_ref().unwrap().attributes.clone()
+        } else {
+            connection_attributes.as_ref().unwrap().clone()
+        };
 
+        // Only attributes that are Either or After are synced to the attributes in the mongo_connection
```
Comment (terakilobyte) [inline odbc/src/api/functions.rs:1365]: oops, dead comment and will be removed next commit!

File: odbc/src/api/functions.rs [hunk 5]
```diff
@@ -2761,9 +2804,49 @@ unsafe fn set_connect_attrw_helper(
     let sql_return = {
         let conn = must_be_valid!((*conn_handle).as_connection());
 
+        let has_connection = conn.has_mongo_connection();
+
+        // Attributes could be set before or after the mongo_connection is established, so we
+        // have to check and write to the proper place. Keep in mind the conn.attributes Option<RwLock<ConnectionAttributes>> is TAKEN
+        // when a mongo_connection is established, so writing to both isn't safe
```
Comment (terakilobyte) [inline odbc/src/api/functions.rs:2811]: Dead comments that will be removed. We discussed offline that we won't bother validating if the dm is passing attributes to us when it should or shouldn't.

File: odbc/src/api/functions.rs [hunk 6]
```diff
@@ -2761,9 +2804,49 @@ unsafe fn set_connect_attrw_helper(
...
+                        .unwrap()
+                        .attributes
+                        .current_catalog = Some(db);
+                } else {
+                    conn.attributes
+                        .write()
+                        .unwrap()
+                        .as_mut()
+                        .unwrap()
+                        .current_catalog = Some(db);
```
Comment (terakilobyte) [inline odbc/src/api/functions.rs:2848]: Thinking about ripping this out into a function.

Comment (terakilobyte) [reply odbc/src/api/functions.rs:2827]: Actually, this would be a bit harder without a wrapper to generalize it.

Comment (pmeredit) [reply odbc/src/api/functions.rs:2827]: It's fine.

File: odbc/src/api/functions.rs [hunk 7]
```diff
@@ -1211,6 +1213,20 @@ fn sql_free_handle(handle_type: HandleType, handle: *mut MongoHandle) -> Result<
                     .as_connection()
                     .ok_or(ODBCError::InvalidHandleType(HANDLE_MUST_BE_CONN_ERROR))?
             };
+            // We must take the connection attributes in the mongo_connection and put them back
+            // in the connection attributes for the conn
+            if conn.has_mongo_connection() {
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:1218]: You shouldn't need to copy these just to drop them.

**Review by pmeredit — CHANGES_REQUESTED** [review]

**Comment by terakilobyte** [comment]
@pmeredit I hit this with a big test. The good news is, it covers all of the movement back and forth between potential locations the attributes could be in depending on the connection state.

File: odbc/src/api/functions.rs [hunk 8]
```diff
@@ -836,6 +836,19 @@ pub unsafe extern "C" fn SQLDisconnect(connection_handle: HDbc) -> SqlReturn {
         || {
             let conn_handle = MongoHandleRef::from(connection_handle);
             let conn = must_be_valid!((*conn_handle).as_connection());
+
+            // prior to dropping the mongo_connection, we need to move it's ConnectionAttributes
+            // to the conn's ConnectionAttributes - the inverse of what we do when a mongo_connection
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:841]:
```suggestion
            // to the conn_handle's ConnectionAttributes - the inverse of what we do when a mongo_connection
```

File: odbc/src/api/functions.rs [hunk 9]
```diff
@@ -1211,6 +1226,20 @@ fn sql_free_handle(handle_type: HandleType, handle: *mut MongoHandle) -> Result<
...
+                let mongo_connection_attributes = conn
+                    .mongo_connection
+                    .read()
+                    .unwrap()
+                    .as_ref()
+                    .unwrap()
+                    .attributes
+                    .clone();
+                let mut connection_attributes = conn.attributes.write().unwrap();
+                *connection_attributes = Some(mongo_connection_attributes);
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:1241]:
```suggestion
                let mongo_connection_attributes = conn
                    .mongo_connection
                    .read()
                    .unwrap()
                    .as_ref()
                    .unwrap()
                    .attributes
                    .clone();
                let mut connection_attributes = conn.attributes.write().unwrap();
                *connection_attributes = Some(mongo_connection_attributes);
```

This clone isn't necessary.

File: odbc/src/api/functions.rs [hunk 10]
```diff
@@ -1211,6 +1226,20 @@ fn sql_free_handle(handle_type: HandleType, handle: *mut MongoHandle) -> Result<
                     .as_connection()
                     .ok_or(ODBCError::InvalidHandleType(HANDLE_MUST_BE_CONN_ERROR))?
             };
+            // We must take the connection attributes in the mongo_connection and put them back
+            // in the connection attributes for the conn
+            if conn.has_mongo_connection() {
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:1231]:
```suggestion
```

Comment (pmeredit) [reply odbc/src/api/functions.rs:1231]: Sorry this is two diffs because github doesn't seem to let you edit around your own comment lol

**Review by pmeredit — CHANGES_REQUESTED** [review]
Feel free to merge after making these changes without retagging me

File: odbc/src/api/functions.rs (hunk 9 above)
Comment (terakilobyte) [reply odbc/src/api/functions.rs]: this is dead code I thought I had removed. It's gone now!

**Comment by terakilobyte** [comment]
evergreen merge

**Pull request merged**